### PR TITLE
Windows build: Add DLL build targets

### PR DESCRIPTION
**Pull request by kebby**
This PR adds two build targets to the Windows build to compile the SDK as a dynamic library instead of  a static one, and a batch file to build a binary distribution from the DLL Release build.

As to why - while binary compatibility under Windows is pretty much perfect, maintaining static libraries can become a hellish nightmare. If a static library doesn't use the exact same platform tool set, SDK version and build flavor as whatever tries to link to it, you're in for a lot of weird and subtle bugs. This means at least having to maintain one version of the SDK per build target, and  can also mean having to recompile the SDK when there's an update to any part of the toolchain, and all kinds of other maintenance shenanigans.

Thus, I added build configurations that compile the CDI SDK as a DLL which, when you add libfabric and the aws-cpp-sdk dlls, only has dependencies that are built into Windows 10, and are stable. Building the SDK is now only necessary when the SDK itself is updated.

Also this helps with runtime performance in case CDI isn't used - we use delayed loading for most 3rd party integrations so unused features won't add to application startup time.

In detail there are the following changes:
- Changed the #ifdef in cdi_utility_api.h to _WIN32 with an underscore because that one's a compiler built-in, and WIN32 wasn't set
- Added CDI_INTERFACE to all public API declarations where it was missing, plus those internal APIs that are used by cdi_test and the examples
- Added explicit dependency from cdi_sdk to libfabric in the project
- Added Release_DLL and Debug_DLL build targets to the solution and cdi_* projects. cdi_sdk will be compiled as DLL in those configurations. pdcurses and libfabric aren't changed.
- Added custom build step (for the _DLL targets) to copy all libfabric related files to the correct output directory so debugging works

As a bonus (I can take those things out of the PR if you want to)
- wrote a deployment script (proj/make_distrib.bat) that copies all necessary files from the Release_DLL config (bin, include, lib, docs) into a target directory (build/aws-cdi-sdk) for distribution in binary form
- turned on multi processor compilation for all cdi_* projects (really helps)
- cleaned up the headers section in the VC project, and the VC related parts of .gitignore

It would be appreciated to get this PR into mainline because it really helps with our own workflow. We're using a lot of 3rd party libraries and integrations, and not having to maintain them build wise unless there's a necessary update is a lot of time saved.

### CODE REVIEW

File: include/cdi_baseline_profile_api.h [hunk 1]
```diff
@@ -270,7 +270,7 @@ CDI_INTERFACE const EnumStringKey* CdiAvmKeyGetArray(CdiAvmBaselineEnumStringKey
  *
  * @return true if the conversion was successful, false if a failure was encountered.
  */
-bool CdiAvmParseBaselineVersionString(const char* version_str, CdiAvmBaselineProfileVersion* ret_version_ptr);
+CDI_INTERFACE bool CdiAvmParseBaselineVersionString(const char* version_str, CdiAvmBaselineProfileVersion* ret_version_ptr);
```
Comment (trevorheaws) [inline include/cdi_baseline_profile_api.h]: We have cut an internal ticket to add a project style guide to our project. In the mean time we limit column width to 120 characters. So please split this onto a second line.

File: include/cdi_logger_api.h [hunk 2]
```diff
@@ -175,7 +176,7 @@ bool CdiLoggerCreateLog(CdiLoggerHandle logger_handle, CdiConnectionHandle con_h
  *
  * @return  true successfully initialized; false if not.
  */
-bool CdiLoggerCreateFileLog(CdiLoggerHandle logger_handle, const char* filename_str, CdiLogHandle* ret_log_handle_ptr);
+CDI_INTERFACE bool CdiLoggerCreateFileLog(CdiLoggerHandle logger_handle, const char* filename_str, CdiLogHandle* ret_log_handle_ptr);
```
Comment (trevorheaws) [inline include/cdi_logger_api.h]: Split to a second line to keep line length below 120 characters.

File: include/cdi_logger_api.h [hunk 3]
```diff
@@ -236,8 +237,8 @@ CdiLogHandle CdiLoggerThreadLogGet(void);
  * @param line_number Source code line number the log message is being generated from.
  * @param state_ptr Pointer to address where to write returned multiline state data.
  */
-void CdiLoggerMultilineBegin(CdiLogHandle log_handle, CdiLogComponent component, CdiLogLevel log_level,
-                             const char* function_name_str, int line_number, CdiLogMultilineState* state_ptr);
+CDI_INTERFACE void CdiLoggerMultilineBegin(CdiLogHandle log_handle, CdiLogComponent component, CdiLogLevel log_level,
+                                           const char* function_name_str, int line_number, CdiLogMultilineState* state_ptr);
```
Comment (trevorheaws) [inline include/cdi_logger_api.h]: line length again.

File: include/cdi_os_api.h [hunk 4]
```diff
@@ -336,8 +338,8 @@ bool CdiOsSignalHandlerSet(int signal_num, SignalHandlerFunction func_ptr);
  *
  * @return true if successful, otherwise false.
  */
-bool CdiOsThreadCreatePinned(ThreadFuncName thread_func, CdiThreadID* thread_id_out_ptr, const char* thread_name_str,
-                             void* thread_func_arg_ptr, CdiSignalType start_signal, int cpu_affinity);
+CDI_INTERFACE bool CdiOsThreadCreatePinned(ThreadFuncName thread_func, CdiThreadID* thread_id_out_ptr, const char* thread_name_str,
```
Comment (trevorheaws) [inline include/cdi_os_api.h]: Split lines to avoid going over 120 characters per line.

File: include/cdi_os_api.h [hunk 5]
```diff
@@ -593,8 +595,8 @@ bool CdiOsSignalWait(CdiSignalType signal_handle, uint32_t timeout_in_ms, bool*
  *
  * @return true if successful, otherwise false.
  */
-bool CdiOsSignalsWait(CdiSignalType* signal_array, uint8_t num_signals, bool wait_all, uint32_t timeout_in_ms,
-                      uint32_t* ret_signal_index_ptr);
+CDI_INTERFACE bool CdiOsSignalsWait(CdiSignalType* signal_array, uint8_t num_signals, bool wait_all, uint32_t timeout_in_ms,
```
Comment (trevorheaws) [inline include/cdi_os_api.h]: Adjust lines so that they stay under 120 characters.

File: include/cdi_os_api.h [hunk 6]
```diff
@@ -737,8 +739,8 @@ bool CdiOsFSeek(CdiFileID file_handle, int64_t offset, int position);
  *
  * @return true if successful, otherwise false.
  */
-bool CdiOsSplitPath(const char* filepath_str, char* filename_str, int filename_buf_size, char* directory_str,
-                    int directory_buf_size);
+CDI_INTERFACE bool CdiOsSplitPath(const char* filepath_str, char* filename_str, int filename_buf_size, char* directory_str,
```
Comment (trevorheaws) [inline include/cdi_os_api.h:743]: Line length

File: include/cdi_pool_api.h [hunk 7]
```diff
@@ -158,8 +160,8 @@ uint32_t CdiPoolGetSizeNeeded(uint32_t item_count, uint32_t item_byte_size);
  *
  * @return true if successful, otherwise false (not enough memory).
  */
-bool CdiPoolCreate(const char* name_str, uint32_t item_count, uint32_t grow_count, uint32_t max_grow_count,
-                   uint32_t item_byte_size, bool thread_safe, CdiPoolHandle* ret_handle_ptr);
+CDI_INTERFACE bool CdiPoolCreate(const char* name_str, uint32_t item_count, uint32_t grow_count, uint32_t max_grow_count,
```
Comment (trevorheaws) [inline include/cdi_pool_api.h:163]: Keep line length under 120.

File: include/cdi_pool_api.h [hunk 8]
```diff
@@ -241,10 +243,10 @@ bool CdiPoolCreateUsingExistingBuffer(const char* name_str, uint32_t item_count,
  *  number of bytes needed in order to create the pool. If true is returned, the value returned in
  *  buffer_byte_size_needed will be the number of bytes actually used for the pool.
  */
-bool CdiPoolCreateUsingExistingBufferAndInitItems(const char* name_str, uint32_t item_count, uint32_t item_byte_size,
-                                                  bool thread_safe, void* buffer_ptr, uint32_t buffer_byte_size,
-                                                  uint32_t* buffer_byte_size_needed_ptr, CdiPoolHandle* ret_handle_ptr,
-                                                  CdiPoolItemOperatorFunction init_fn, void* init_context_ptr);
+CDI_INTERFACE bool CdiPoolCreateUsingExistingBufferAndInitItems(const char* name_str, uint32_t item_count, uint32_t item_byte_size,
```
Comment (trevorheaws) [inline include/cdi_pool_api.h:246]: Line length again. This function name is so long it definitely makes this area a bit clunky but still would like to keep to the line length.

File: include/cdi_pool_api.h [hunk 9]
```diff
@@ -362,7 +364,7 @@ int CdiPoolGetFreeItemCount(CdiPoolHandle handle);
  * @return false if any items are currently allocated from the pool or if operator_function returned false for at least
  *         one item in the pool, otherwise true.
  */
-bool CdiPoolForEachItem(CdiPoolHandle handle, CdiPoolItemOperatorFunction operator_function, const void* context_ptr);
+CDI_INTERFACE bool CdiPoolForEachItem(CdiPoolHandle handle, CdiPoolItemOperatorFunction operator_function, const void* context_ptr);
```
Comment (trevorheaws) [inline include/cdi_pool_api.h:367]: Line length again.

File: include/cdi_queue_api.h [hunk 10]
```diff
@@ -77,8 +77,8 @@ extern "C" {
  *
  * @return true if successful, otherwise false is returned.
  */
-bool CdiQueueCreate(const char* name_str, uint32_t item_count, uint32_t grow_count, uint32_t max_grow_count,
-                    uint32_t item_byte_size, CdiQueueSignalMode signal_mode, CdiQueueHandle* ret_handle_ptr);
+CDI_INTERFACE bool CdiQueueCreate(const char* name_str, uint32_t item_count, uint32_t grow_count, uint32_t max_grow_count,
```
Comment (trevorheaws) [inline include/cdi_queue_api.h:80]: Make line length less than or equal to 120 characters per line.

File: include/cdi_queue_api.h [hunk 11]
```diff
@@ -105,7 +105,7 @@ bool CdiQueuePop(CdiQueueHandle handle, void* item_dest_ptr);
  *
  * @return true if successful, otherwise false (queue is empty and timeout expired or signal got set).
  */
-bool CdiQueuePopWait(CdiQueueHandle handle, int timeout_ms, CdiSignalType abort_wait_signal, void* item_dest_ptr);
+CDI_INTERFACE bool CdiQueuePopWait(CdiQueueHandle handle, int timeout_ms, CdiSignalType abort_wait_signal, void* item_dest_ptr);
```
Comment (trevorheaws) [inline include/cdi_queue_api.h:108]: Line length again.

File: include/cdi_queue_api.h [hunk 12]
```diff
@@ -148,7 +148,7 @@ bool CdiQueuePush(CdiQueueHandle handle, const void* item_ptr);
  *
  * @return true if successful, otherwise false (queue is full and timeout expired or signal got set).
  */
-bool CdiQueuePushWait(CdiQueueHandle handle, int timeout_ms, CdiSignalType abort_wait_signal, const void* item_ptr);
+CDI_INTERFACE bool CdiQueuePushWait(CdiQueueHandle handle, int timeout_ms, CdiSignalType abort_wait_signal, const void* item_ptr);
```
Comment (trevorheaws) [inline include/cdi_queue_api.h]: line length

File: include/cdi_queue_api.h [hunk 13]
```diff
@@ -165,8 +165,8 @@ bool CdiQueuePushWait(CdiQueueHandle handle, int timeout_ms, CdiSignalType abort
  *
  * @return true if successful, otherwise false (queue is empty and timeout expired or signal got set).
  */
-bool CdiQueuePushWaitMultiple(CdiQueueHandle handle, int timeout_ms, CdiSignalType* abort_wait_signal_array,
-                              int num_signals, uint32_t* ret_signal_index_ptr, const void* item_ptr);
+CDI_INTERFACE bool CdiQueuePushWaitMultiple(CdiQueueHandle handle, int timeout_ms, CdiSignalType* abort_wait_signal_array,
```
Comment (trevorheaws) [inline include/cdi_queue_api.h]: line length

File: src/common/include/fifo_api.h [hunk 14]
```diff
@@ -85,8 +85,8 @@ typedef void (*CdiFifoCallback)(const CdiFifoCbData* data_ptr);
  *
  * @return true if successful, otherwise false (not enough memory to allocate the FIFO).
  */
-bool CdiFifoCreate(const char* name_str, int item_count, int item_byte_size, CdiFifoFullCallback full_cb_ptr,
-                   CdiUserCbParameter full_user_cb_param, CdiFifoHandle* ret_handle_ptr);
+CDI_INTERFACE bool CdiFifoCreate(const char* name_str, int item_count, int item_byte_size, CdiFifoFullCallback full_cb_ptr,
```
Comment (trevorheaws) [inline src/common/include/fifo_api.h]: line length.

File: src/common/include/fifo_api.h [hunk 15]
```diff
@@ -99,14 +99,14 @@ bool CdiFifoCreate(const char* name_str, int item_count, int item_byte_size, Cdi
  *
  * @return true if successful, otherwise false (FIFO is empty).
  */
-bool CdiFifoRead(CdiFifoHandle handle, int timeout_ms, CdiSignalType abort_wait_signal, void* item_dest_ptr);
+CDI_INTERFACE bool CdiFifoRead(CdiFifoHandle handle, int timeout_ms, CdiSignalType abort_wait_signal, void* item_dest_ptr);
```
Comment (trevorheaws) [inline src/common/include/fifo_api.h]: line length

File: src/common/include/fifo_api.h [hunk 16]
```diff
@@ -118,14 +118,14 @@ void CdiFifoFlush(CdiFifoHandle handle);
  *
  * @return true if successful, otherwise false (FIFO is full).
  */
-bool CdiFifoWrite(CdiFifoHandle handle, int timeout_ms, CdiSignalType abort_wait_signal, const void* item_ptr);
+CDI_INTERFACE bool CdiFifoWrite(CdiFifoHandle handle, int timeout_ms, CdiSignalType abort_wait_signal, const void* item_ptr);
```
Comment (trevorheaws) [inline src/common/include/fifo_api.h]: line length

**Review by trevorheaws — CHANGES_REQUESTED** [review]
Thank you for the submission. I have posted a number of comments flagging spots where line length exceeds 120 characters. I have created an internal task for us to detail our our code formatting conventions in a publicly accessible document. Thank you for your patience on this and the high quality contribution.

**Comment by kebby** [comment]
Ok, line lengths are fixed, and there was also a hiccup with the project file thanks to my git ineptitude but that should be good now, too.

Thanks for considering this PR, this really saves us a lot of work.

**Review by trevorheaws — APPROVED** [review]
Thank you for making the line length changes.

**Pull request merged**